compiler/src/iree/compiler/Codegen/Common/GPU/AMDGPUDistributeContract.cpp

```diff
@@ -0,0 +1,269 @@
+                                       NestedLayoutAttr rhsLayout) const {
+    int64_t lhsKBatch = 0, rhsKBatch = 0;
+    if (kind == ContractKind::MK_KN_MN) {
+      lhsKBatch = lhsLayout.getBatchesPerSubgroup()[1];
+      rhsKBatch = rhsLayout.getBatchesPerSubgroup()[0];
+    } else {
+      return std::nullopt;
+    }
+
+    if (lhsKBatch != rhsKBatch)
+      return std::nullopt;
+    return lhsKBatch;
+  }
+
+  // Given a contract op's batch |resultOffsets|, gets its batch offsets for
+  // both LHS and RHS.
+  bool getOperandBatchOffsets(
+      ContractKind kind, int64_t kOffset,
+      // Copy intentionally given we need to mutate in the function body
+      SmallVector<int64_t, 2> resultOffsets, NestedLayoutAttr resultLayout,
+      SmallVector<int64_t, 2> &lhsOffsets, SmallVector<int64_t, 2> &rhsOffsets,
+      NestedLayoutAttr lhsLayout, NestedLayoutAttr rhsLayout) const {
+    // The result offsets are permutated and we need to revert the permutation.
+    // The following works based on the fact that for 2-D cases, a permuation
+    // vector's reverse is just itself. So this only works for 2-D cases.
+    applyPermutationToVector(resultOffsets, resultLayout.getBatchOrder());
+
+    // resultOffsets contains batch indices into the C/D vector. It is a 2-D
+    // index for both M and N. We need to split out for M and N, and add index
+    // for K.
+    if (kind == ContractKind::MK_KN_MN) {
+      lhsOffsets[0] = resultOffsets[0];
+      lhsOffsets[1] = kOffset;
+      rhsOffsets[0] = kOffset;
+      rhsOffsets[1] = resultOffsets[1];
+    } else {
+      return false;
+    }
+
+    // Now apply permutation on lhs/rhs according to their batch order.
+    applyPermutationToVector(lhsOffsets, lhsLayout.getBatchOrder());
+    applyPermutationToVector(rhsOffsets, rhsLayout.getBatchOrder());
+    return true;
+  }
+
+  // Returns the vector type for the given |value| meeting mfma ops's
+  // requirements.
+  VectorType getMFMAVectorType(Value value) const {
+    auto type = cast<VectorType>(value.getType());
+    SmallVector<int64_t> shape;
+    for (int64_t dim : type.getShape()) {
+      if (dim != 1)
+        shape.push_back(dim);
+    }
+    return VectorType::get(shape, type.getElementType());
+  }
+
+  Value computeMMA(OpBuilder &builder, Location loc, Value a, Value b, Value c,
+                   VectorType aType, VectorType bType, VectorType cType,
+                   IREE::GPU::MFMAIntrinsic intrinsic) const {
+    Value aCast = builder.create<vector::ShapeCastOp>(a.getLoc(), aType, a);
+    Value bCast = builder.create<vector::ShapeCastOp>(b.getLoc(), bType, b);
+    Value cCast = builder.create<vector::ShapeCastOp>(c.getLoc(), cType, c);
+
+    uint32_t m, n, k, blocks;
+    switch (intrinsic) {
+    case IREE::GPU::MFMAIntrinsic::F16_16x16x16_F32:
+      m = n = k = 16;
+      blocks = 1;
+      break;
+    case IREE::GPU::MFMAIntrinsic::F16_32x32x8_F32:
+      m = n = 32;
+      k = 8;
+      blocks = 1;
+      break;
```

Comment:
nit: add `llvm_unreachable` default for NYI cases

Reply:
Should be fine here given `MFMAIntrinsic` is an `enum class` so if something is missing compiler will yell at us.

Reply:
oh I didn't realize that, that's cool